### PR TITLE
Ml for synthesis prototype - Refactoring

### PR DESCRIPTION
**Pull request by ryukzak**
Changes overview:

- Add Makefile with key project actions (build, test, lint, format).
- Reorganise CI flow. Separate ml and ui jobs from nitta build/test.
- Add popover with all scores to subforest screen.
- Bump stackage snapshot, to be hls-powered.
- Some haskell related refactoring.
- Update tensorflow from 2.12 to 2.13. Add mac os specific dependencies.
- Add hack to specify `NITTA_RUN_COMMAND` from env variable.

TODOs:

- [ ] Docker image for development: add make target to multi-platform build and push docker to docker-hub
- [ ] Replace black via ruff autoformatter in CI, Makefile, and dev-docker-image
- [ ] Rename `score` field in `NodeView` to `defScore` (UI, Haskell, maybe ML backend)
- [ ] Add score selector to subForest screen.
- [ ] Change CLI argument `port` type from `Int` with magic `-1` to `Maybe Int`
- [ ] Push ML scores to STM during synthesis process (`NITTA.Synthesis.Explore:subForestIO`).

```haskell
subForestIO
    BackendCtx{nodeScores, mlBackendGetter}
    tree@Tree{sSubForestVar} = do
        (firstTime, subForest) <-
            atomically $
                tryReadTMVar sSubForestVar >>= \case
                    Just subForest -> return (False, subForest)
                    Nothing -> do
                        subForest <- exploreSubForestVar tree
                        putTMVar sSubForestVar subForest
                        return (True, subForest)

        when firstTime $ traceProcessedNode tree

        -- FIXME: ML scores are evaluated here every time subForestIO is called. how to cache it like the default score? IO in STM isn't possible.
        -- also it looks inelegant, is there a way to refactor it?
        let modelNames = mapMaybe (T.stripPrefix mlScoreKeyPrefix) nodeScores
        if
            | null subForest -> return subForest
            | null nodeScores -> return subForest
            | null modelNames -> return subForest
            | otherwise -> do
                MlBackendServer{baseUrl} <- mlBackendGetter
                case baseUrl of
                    Nothing -> return subForest
                    Just mlBackendBaseUrl -> do
                        -- (addMlScoreToSubforestSkipErrorsIO subForestAccum modelName) gets called for each modelName
                        foldM (addMlScoreToSubforestSkipErrorsIO mlBackendBaseUrl) subForest modelNames
```

- [ ] Dockerfile with dev containers and other related stuff (vscode workspace) should be relocated from ml/synthesis to a separate dir, perhaps.
- [ ] Currently ML if the config is invalid, the new evaluate script will fail with generic errors (TypeError, KeyError, etc.). It would be useful to write a JSON Schema describing evaluation configs. It will help to provide relevant field descriptions in editors and to implement a config validation in the script itself.
- [ ] Simplify environment setup and nitta -- MLBackend integration. It shouldn't have such complex, like it happens in `Makefile:ml-nitta`
- [ ] During the model training, if we use completely different synthesis trees for validation dataset instead of a random split over all shuffled training data, the validation loss and mae will be significantly worse. It means that the current model doesn't generalize well to other synthesis trees, although synthesis may still succeed, as it was shown. Improving those new validation metrics is crucial for further ML synthesis development. Need more input node parameters? Adjusting random descent in MCTS-like style too to fix negative labels share? UPD: some related work was done and yielded significant results in regards to fixing this, but mode's generalization abilities still require improvement.
- [ ] Add tensorflow version check during dev-image build (should be the same with pyproject.toml)
- [ ] I think the best solution here is to add `NITTA_ML_BACKEND_RUN_COMMAND` to provide more control. But I move it to the new issue because:

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+ml-lint:
+	$(POETRY) run ruff $(PYTHONPATH)
+	$(POETRY) run mypy --config-file ml/synthesis/pyproject.toml $(PYTHONPATH)
+	cd $(POETRYPATH) && poetry run vulture
+
+ml-nitta:
+	echo 'Model:' $(ML_MODEL)
+	$(POETRY) shell
+	MODELS_DIR=$(ML_MODEL_PATH) PYTHONPATH=$(PYTHONPATH) stack exec nitta -- examples/teacup.lua -s ml_$(ML_MODEL) -p=8080
```
Comment (iburakov) [inline Makefile:95]: Actually, it isn't too scary already, is it?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,115 @@
+HS_SRC_DIR = app src test
+HS_O = --fast
+
+POETRYPATH = ml/synthesis
+PYTHONPATH = ml/synthesis/src
+POETRY = poetry -C $(POETRYPATH)
+PYTHON = PYTHONPATH=$(PYTHONPATH) $(POETRY) run python3
+ML_MODEL_PATH = ml/synthesis/models
+ML_MODEL = $(shell ls -t ml/synthesis/models | grep model | head -n 1)
```
Comment (iburakov) [inline Makefile:9]: ml/synthesis/models -> ML_MODEL_PATH?

**Review by iburakov — CHANGES_REQUESTED** [review]
Mostly minor stuff, yet important enough.

Suggested new TODOs:
- [ ] write all `make docker-dev-*` commands and update the ml/synthesis/README.md

File: Makefile (hunk 1 above)
Comment (iburakov) [reply Makefile:95]: I've given some more thought to the python environment issue.

Well, the obvious/purist solution is to always keep the development environment right (so `python` always resolves to the correct one) and take it for granted in the scripts.

It's what's done in the dev container, it's what IDEs do: VS Code and PyCharm activate the project venv by default for all terminal sessions and set the correct PYTHONPATH, so everything just works. This solution is battle-tested and convenient... unless it isn't, I guess :)

I think, it became a problem for you because you look from a perspective that's different from mine: you just need to run mlbackend and that's it. I.e. you don't care about the active python environment and current PYTHONPATH and don't want to mess with it at all (and that's sane!).

Comment (iburakov) [reply Makefile:95]: What's the assumption `nitta` makes when you run the ml scoring? That `python -m mlbackend` will do the right thing. It was convenient for me since I've always had the right python environment set up (both locally and in the dev container), but it may not be the case for everyone in all circumstances, I guess.

So, I think the simplest way is to, indeed, write a slightly more clever logic on the Haskell side to find the `mlbackend` to run. It may try to invoke `poetry run python -m mlbackend` prior to calling the system interpreter, for example.

In the "distributable NITTA version" for end users it may just look for a single `mlbackend.exe`. I've tried to bundle it this way once (with pyInstaller, iirc) and it worked!

Just the NITTA_MLBACKEND_RUN_COMMAND seems like a fine solution too, but that's another env var to set for developers :)

There are also some crazy ideas like "install a shim module called mlbackend to the system python interpreter that just finds the correct venv and runs the correct module", but that's way too overcomplicated. And still another python package to install to the system interpreter for developers.

File: ml/synthesis/src/components/data_crawling/nitta/nitta_running.py [hunk 3]
```diff
@@ -144,6 +144,10 @@ async def run_nitta_server(
 ) -> AsyncGenerator[NittaRunResult, None]:
     port = 0  # NITTA will choose a random free port and print it to stdout, we'll parse
 
+    if new_nitta_cmd := os.environ.get("NITTA_RUN_COMMAND"):
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply ml/synthesis/src/components/data_crawling/nitta/nitta_running.py:147]: `consts.NITTA_RUN_COMMAND_OVERRIDE` -- is better, because we have a lot of source of nitta command, including configuration files.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+ui-format-check:
+	yarn --cwd web exec -s prettier -- --check 'web/src/**/*.{ts,tsx}' --ignore-path web/.gitignore
+
+############################################################
+## nitta ml
+############################################################
+
+ml-crawl-data:
+	$(PYTHON) $(PYTHONPATH)/scripts/crawl_data_by_tree_sampling_many.py
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply Makefile:76]: Don't sure which option is better. First -- more explicit. Second -- more compact and pythonic. Let use yours.

File: ml/synthesis/src/scripts/evaluate_nitta_synthesis.py [hunk 5]
```diff
@@ -280,7 +280,7 @@ def _aggregate_and_save_results(results: list[dict], config: EvaluationConfig):
     if not results:
         return
 
-    config.output_dir.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply ml/synthesis/src/scripts/evaluate_nitta_synthesis.py:283]: In some case I receive str object here and catch run time error. So I fix it in a hack way, instead of a debugging.

File: src/NITTA/Synthesis/Method.hs [hunk 6]
```diff
@@ -218,11 +216,6 @@ topDownByScoreSynthesisIO' heap step depthCoeffBase limit scoreKey ctx currentNo
 
                             topDownByScoreSynthesisIO' (H.drop dropCount heapWithSubforest) (step + 1) depthCoeffBase limit scoreKey ctx nextBestScoreNode
 
-{- | Shortcut for constraints in signatures of synthesis method functions.
-This used to be (VarValTime v x t, UnitTag tag). See below for more info.
--}
-type SynthesisMethodConstraints tag v x t = (VarValTimeJSON v x t, ToJSON tag, UnitTag tag)
-
 -- FIXME: Validate the type above, its usages and meaning in the context of changes described below.
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply src/NITTA/Synthesis/Method.hs:219]: I don't sure is I answer to this question or not, but the main purpose of explicit type signatures here is: https://ryukzak.github.io/nitta/haddock/nitta/NITTA-Synthesis-Method.html

Without type alias it transform in a mess.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+############################################################
+## docker development image
+############################################################
+
+docker-dev-build:
+	docker build \
+		--target development \
+		-f ml/synthesis/Dockerfile \
+		-t nitta-dev \
+		.
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply Makefile:106]: Owww, it is painful....

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	$(POETRY) run mypy --config-file ml/synthesis/pyproject.toml $(PYTHONPATH)
+	cd $(POETRYPATH) && poetry run vulture
+
+ml-nitta:
+	echo 'Model:' $(ML_MODEL)
+	$(POETRY) shell
+	MODELS_DIR=$(ML_MODEL_PATH) PYTHONPATH=$(PYTHONPATH) stack exec nitta -- examples/teacup.lua -s ml_$(ML_MODEL) -p=8080
+
+############################################################
+## docker development image
```
[the remark this answers is not in the document]
Comment (ryukzak) [reply Makefile:98]: It is very tricky for mac os right now, because it have M1 and Intel processors at the same time. So just to make simpler -- we can forget for GPU on mac.

I make it by conditional statement (win/linux/mac). But for GPU -- explicitly.

File: Makefile (hunk 1 above)
Comment (ryukzak) [reply Makefile:95]: >  I've tried to bundle it this way once (with pyInstaller, iirc) and it worked!

Miracle?

I think the best solution here is to add `NITTA_MLBACKEND_RUN_COMMAND` to provide more control. But I move it to the new issue because:

```haskell
    -- not a shell process, because killing ML backend under shell is trickier
    -- let's find out python3 executable path
```

File: ml/synthesis/src/consts.py [hunk 9]
```diff
@@ -49,3 +49,6 @@ def _find_root_dir():
 MODELS_DIR = Path(_models_dir_env) if _models_dir_env else ML_SYNTHESIS_DIR / "models"
 
 ML_BACKEND_BASE_URL_FILEPATH = ".ml_backend_base_url"
+
+# high priority env var which overrides command provided in the config file
+NITTA_RUN_COMMAND_OVERRIDE = os.environ.get("NITTA_RUN_COMMAND", None)
```
Comment (iburakov) [inline ml/synthesis/src/consts.py:54]: used env var name (`NITTA_RUN_COMMAND`) doesn't match
https://github.com/ryukzak/nitta/blob/ecd35f948ec03b9920b8f3fd06e73207bc8ef6ca/.github/workflows/ci.yml#L246-L247

Comment (iburakov) [reply ml/synthesis/src/consts.py:54]: also it may be better to do it like
```python3
class EnvVarNames:
    MODELS_DIR = "NITTA_ML_SYNTHESIS_MODELS_DIR"
    NITTA_RUN_COMMAND = "NITTA_RUN_COMMAND"

...

NITTA_RUN_COMMAND_OVERRIDE = os.environ.get(EnvVarNames.NITTA_RUN_COMMAND, None)
```

it's consistent with the 
```python3
_models_dir_env = os.environ.get(EnvVarNames.MODELS_DIR)
```

File: Makefile [hunk 10]
```diff
@@ -90,21 +94,60 @@ ml-lint:
...
 ############################################################
 
-docker-dev-build:
+
+docker-dev-build-for-linux-win:
+	docker build \
+		--target development \
+		-f ml/synthesis/Dockerfile \
+		--build-arg HOST_UID=$(id -u) \
+		--build-arg HOST_GID=$(id -g) \
```
Comment (iburakov) [inline Makefile:114]:
```suggestion
		--build-arg HOST_UID=$(shell id -u) \
		--build-arg HOST_GID=$(shell id -g) \
```

otherwise it's 
```shell
        --build-arg HOST_UID= \
        --build-arg HOST_GID= \
```

File: Makefile [hunk 11]
```diff
@@ -90,21 +94,60 @@ ml-lint:
...
+	make docker-dev-build-for-linux-win
+endif
+
+
+docker-dev-build-with-gpu-for-linux-win:
+	docker build \
+		--target development-gpu \
+		-f ml/synthesis/Dockerfile \
+		--build-arg HOST_UID=$(id -u) \
+		--build-arg HOST_GID=$(id -g) \
```
Comment (iburakov) [inline Makefile:139]: same as above,
```suggestion
		--build-arg HOST_UID=$(shell id -u) \
		--build-arg HOST_GID=$(shell id -g) \
```

File: ml/synthesis/pyproject.toml [hunk 12]
```diff
@@ -19,7 +19,11 @@ uvicorn = { extras = ["standard"], version = "^0.23.2" }
 pandas = { version = "2.0.2", extras = ["performance"] }
 tabulate = "^0.9.0"
 
-tensorflow = {version = "^2.13.1" }
+# source: https://github.com/python-poetry/poetry/issues/8271#issuecomment-1696053565
```
Comment (iburakov) [inline ml/synthesis/pyproject.toml:22]: doesn't work on windows, needs fixing, will do myself, resolve when done

Comment (iburakov) [reply ml/synthesis/pyproject.toml:22]: and poetry.lock needs updating

**Review by iburakov — CHANGES_REQUESTED** [review]
Almost there

File: ml/synthesis/pyproject.toml (hunk 12 above)
Comment (iburakov) [reply ml/synthesis/pyproject.toml:22]: done in #261

File: ml/synthesis/src/consts.py (hunk 9 above)
Comment (ryukzak) [reply ml/synthesis/src/consts.py:54]: Don't fully understand the idea behind `EnvVarNames` class, but consistency is an argument. Thank you!

Comment (iburakov) [reply ml/synthesis/src/consts.py]: The idea is quite boring, EnvVarNames.MODELS_DIR had to be reused in tests: 
https://github.com/ryukzak/nitta/blob/ffa93028f8f30ab34b84643dc658cbfe66c0dca9/ml/synthesis/src/tests/test_smoke.py#L80-L86

So I preferred to do it this way instead of hardcoding :)
Class is just for namespacing.

Regardless, I think it's good to have a single place where all the env var names (that python modules understand and use) are listed together, so they're not scattered across the code as literals. If there ever will be a need to generate docs for them, it'll make it simpler.

File: Makefile (hunk 10 above)
Comment (iburakov) [reply Makefile:114]: Missed this one

File: src/NITTA/Synthesis/Method.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iburakov) [reply src/NITTA/Synthesis/Method.hs]: I see, got it. Yes, it's an answer, thank you.

It's a shame it caused me to spend another couple of hours on debugging when ToJSON constraints had to be added. Compiler errors looked completely cryptic to a Haskell junior that was completely unaware those constraints exist :D

File: ml/synthesis/src/scripts/evaluate_nitta_synthesis.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (iburakov) [reply ml/synthesis/src/scripts/evaluate_nitta_synthesis.py:283]: Oh, yes, right, it's probably a bug in EvaluationConfig loading introduced during a stdlib-only rewrite. 

Should be fixed anyway, config.output_dir must be a Path (type hints say so, it's disorienting if it's suddenly a str). But alright, later, perhaps. Should we make it new little issue?

**Review by iburakov — APPROVED** [review]
